### ISSUE TITLE
Bag of Tricks for Image Classification with Convolutional Neural Networks

### ISSUE THREAD

**Issue by Kaniikura** [Kaniikura/paper-survey#14]
## 一言でいうと
CNNによる画像分類における，前処理や訓練プロセスの
様々なテクニックについて実験的にその有効性を確かめた．
### 論文リンク
https://arxiv.org/abs/1812.01187
### 著者/所属機関
Tong He, Zhi Zhang, Hang Zhang, Zhongyue Zhang, Junyuan Xie, Mu Li
* Amazon Web Services
### 投稿日付(yyyy/MM/dd)
2018/12/05
## 新規性・差分
CNNの論文ではモデルアーキテクチャの説明の要点がおかれ，
学習時の細かなトリックについて述べることは稀である．
本論文では，実装のソースコードで確認できる学習プロセスの工夫や
モデルアーキテクチャの微調整など性能向上のための
トリックについて網羅的な調査を行った．<br>
有効性の確認された手法を組み合わせて，既存モデルを学習させたところ
性能が大幅に向上し最新モデルを超えることができた．
### ベースライン
ResNet-50, InceptionV3およびMobileNetの３モデルについて，
それぞれ[ResNetでよく使用される前処理やバッチサイズ，学習率の設定](http://torch.ch/blog/2016/02/04/resnets.html)を
用いたものをベースラインとした．
<img src=https://user-images.githubusercontent.com/32826608/62641885-8d0dfc80-b97f-11e9-9eb9-9e03a421e10b.png width=60%)>
性能向上の手法は①訓練の効率化，②モデルの微調整および③訓練の洗練化という
３つに大別して検証する．
#### バッチサイズと学習率
* **Linear scaling**
ミニバッチSGDではバッチサイズが増えても勾配の期待値は変わらないが分散は小さくなる．
これは学習のノイズが減ることを意味し，この理由から学習率をバッチサイズに比例させて
増やすアプローチが有効である．
* **LR warmup**
学習のはじめは，モデルのパラメータがランダムに初期化された状態であり
学習がなかなか安定しにくい．そのため学習率を減らして学習を安定化させ，段々と
大きな学習率に戻していくlearning rate warmupという手法がとられることがある．
    e.g. はじめのmバッチをwarmup段階とし，学習率のデフォルト値をηとする．
    このときi番目のバッチにおける学習率はiη/mとなる．
#### バイアス
* **zero γ**
ResNetにおける各blockの出力は入力をxとしてx + block(x)と表せる．
このときBlockの最終層はbatch normalization (BN) layerでありバッチ入力**H**を標準化して
**Hˆ**になおした後，γ**Hˆ** + βという線形変換を施す．このときγを0で初期化すると
次のblockに以前のblock出力が伝わりやすくなり学習の安定性に貢献する．
* **No bias decay**
 L2正則化を全てのパラメータ(weight, bias)に作用させるよりも，weightだけに適用し
biasやBNのパラメータγ, βは対象外としたほうが良い．
#### Low-precision training
*  **FP16**
FP32からFP16にパラメータのprecisionを変更することで，精度を犠牲に学習スピードを
2, 3倍ほど向上させることができる．ただ一方で，精度が下がることにより訓練の進行具合に
影響を及ぼす可能性もある．解決策として例えば勾配はFP16で計算するが，更新用の
モデルパラメータはFP32で別に保存して置いたものを使用するなどの手法がある．
#### 結果
<img src=https://user-images.githubusercontent.com/32826608/62642154-158c9d00-b980-11e9-991a-f1cf5429d729.png width=60%>

### 検証② モデルの微調整
例としてResNetの場合を述べる．
<img src=https://user-images.githubusercontent.com/32826608/62642385-8af86d80-b980-11e9-8ac1-4077cb895677.png width=60%>
ResNetは入力を受け付けるinput stemとそれに続く4つのstageそして出力を行う最終層からなる．<br>
input stemではストライドが2である7×7convと3×3max poolingをかけており，
入力はサイズが1/4にダウンサンプリングされチャネル数が64になる．<br>
それぞれのstageではまずダウンサンプリングが行われ，いくつかのResidual blockが
引き続く形になる．そのダウンサンプリングの方法は特徴的であり，それぞれの畳込み処理が
異なるpathが２つ存在する (上図参照)．結果だけ見れば，このdown sampling blockで
入力はサイズが1/2, チャネル数が4倍になって出力される．
また下図に示すようにdown sampling blockにはBとCのようなバリエーションがあり．
Dは本論文で新たに提案する構造である．
<img src=https://user-images.githubusercontent.com/32826608/62643219-6dc49e80-b982-11e9-9ce6-642133a0243c.png width=60%>
また，residual blockはconvのstrideが1であることを除きdown sampling blockと似た構造である．
stride=1のため入出力で画像のサイズは変わらない．また，ResNetのアーキテクチャは
residual blockの数を変えて様々な種類が得られ，例えばResNet-50やResNet-152の数字は
モデルのconv層の数を表している．
#### 結果
さて，down sampling blockの変更による性能の違いは以下のようになった．
<img src=https://user-images.githubusercontent.com/32826608/62644571-618e1080-b985-11e9-809e-e646db26cd85.png width=60%>

#### Cosine Learning Rate Decay
学習率の調整は訓練進展に伴って減少させる方法が一般的である．中でも，
ステップ数に対して指数関数的もしくは階段状に減少させる方法が広く用いられている．
また["SGDR: Stochastic Gradient Descent with Warm Restarts"](https://arxiv.org/abs/1608.03983)において提案されたcosine annealingは学習率をcos関数的に減少させている．その式はTをバッチ数の合計，tを現在のバッチとするとwarmupを無視して
<p align=center><img src=https://user-images.githubusercontent.com/32826608/62645157-9fd7ff80-b986-11e9-9ac9-3a970ddf28fa.png width=40%></p>
のように表される．下図でみるように階段的な学習率調整と比較して訓練初期と終期における
学習率の変動がゆるやかである．
<img src=https://user-images.githubusercontent.com/32826608/62645619-8c796400-b987-11e9-999c-4df4a44da031.png width=60%>

#### Label Smoothing
クラス分類におけるモデルの予測確率は以下のようにsoftmax関数によって正規化された表現になる．
<p align=center><img src=https://user-images.githubusercontent.com/32826608/62658716-644c2e00-b9a4-11e9-9591-68a108736659.png width=30%></p>
このようなモデルの予測と，確率の真の分布(もとの画像のラベルyについてp_i = 1 if i=y else 0)の交差エントロピーは
<p align=center><img src=https://user-images.githubusercontent.com/32826608/62659160-9316d400-b9a5-11e9-871c-3bff78a7f7d5.png width=30%></p>
でありこれを書き直すと
<p align=center><img src=https://user-images.githubusercontent.com/32826608/62659260-dbce8d00-b9a5-11e9-86e2-5fbeab88b40a.png width=30%></p>
すなわち，softmaxに渡す前のモデル出力が正解ラベルの箇所で無限大になることが
最適解となってしまい，これはover-fitの原因になる．label smoothingは正解ラベルで1,その他は0という鋭いピークをもった確率分布を以下のように緩和した表現にすることである． 
<p align=center><img src=https://user-images.githubusercontent.com/32826608/62659399-37991600-b9a6-11e9-9486-a865544f82d3.png width=30%></p>
この最適解は以下のように有限となるため，おそらくモデルの汎化性能を高めると期待される．
<p align=center><img src=https://user-images.githubusercontent.com/32826608/62659607-cc9c0f00-b9a6-11e9-84ee-7d49799412a6.png width=50%></p>

#### Knowledge Distillation
蒸留とも．事前学習済みで精度の高い教師モデルと軽量な生徒モデルの２つが存在し，
教師モデルが生徒モデルの学習をhelpする．pを真のターゲット確率分布，zとrをそれぞれ
生徒と教師モデルの最終層出力とする．この場合，生徒モデルの誤差関数はzとpの交差エントロピーを考える代わりに以下を用いる．
<p align=center><img src=https://user-images.githubusercontent.com/32826608/62712935-d1a89f00-ba36-11e9-8263-1a16f93d5029.png width=60%></p>
Tは温度パラメータと呼ばれ，softmaxの出力を滑らかにして教師モデルの確率分布出力から情報を引き出しやすいように調整する役割がある．

#### Mixup Training
augmentationの一種．ラベルと画像をある割合で混ぜ合わせて訓練に用いる．
<p align=center><img src=https://user-images.githubusercontent.com/32826608/62713726-2bf62f80-ba38-11e9-86d5-08a8a0471938.png width=50%></p>
教師データの線形補間をデータセットに組み込む形になり，モデルの性能向上に効くときがある．

#### 結果
<img src=https://user-images.githubusercontent.com/32826608/62714186-06b5f100-ba39-11e9-92b9-5f3e2aea00f7.png width=80%>

## コメント
本当は転移学習に適用した結果とか，全部の要素考えてResNet-50をtuningした話もあったが本質的ではないため省略した．
